src/java/org/apache/cassandra/cql3/CqlConstraint.java

```diff
@@ -0,0 +1,151 @@
+    }
+
+    public void appendCqlTo(CqlBuilder builder)
+    {
+        builder.append(" CHECK ").append(toString());
+    }
+
+    public void checkConstraint(Map<String, String> columnValues, ColumnMetadata columnMetadata, TableMetadata tableMetadata)
+    {
+        constraintCondition.checkCondition(columnValues, columnMetadata, tableMetadata);
+    }
+
+    public void validateConstraint(ColumnMetadata columnMetadata, TableMetadata tableMetadata)
+    {
+        constraintCondition.validateCondition(columnMetadata, tableMetadata);
+    }
+
+    @Override
+    public String toString()
+    {
+        return constraintCondition.toString();
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hashCode(constraintName);
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
+        return obj instanceof CqlConstraint
+               && Objects.equal(constraintName, ((CqlConstraint) obj).constraintName)
+               && Objects.equal(constraintCondition, ((CqlConstraint) obj).constraintCondition);
+    }
+
+
+    public static class Serializer
+    {
+        public void serialize(CqlConstraint cqlConstraint, DataOutputPlus out, Version version) throws IOException
+        {
+            out.writeUTF(cqlConstraint.constraintName.toString());
+            out.writeUTF(cqlConstraint.columnName.toString());
+            out.writeUTF(cqlConstraint.constraintCondition.getClass().toString());
+            cqlConstraint.constraintCondition.serialize(cqlConstraint.constraintCondition, out, version);
+        }
+
+        public CqlConstraint deserialize(DataInputPlus in, String keyspace, AbstractType<?> columnType, Types types, UserFunctions functions, Version version) throws IOException
+        {
+            String nameText = in.readUTF();
+            String columnName = in.readUTF();
+            ColumnIdentifier identifier = new ColumnIdentifier(nameText, true);
+            ColumnIdentifier columnNameIdentifier = new ColumnIdentifier(columnName, true);
+            String columnConstraintClassName = in.readUTF();
+            ConstraintCondition condition = ConstraintSerializerFactory.getCqlConditionSerializer(columnConstraintClassName)
+                                                                       .deserialize(in, keyspace, columnType, types, functions, version);
+            return new CqlConstraint(identifier, columnNameIdentifier, condition);
+        }
+    }
+
+    public static class ConstraintSerializerFactory
+    {
+        public static CqlConstraintSerializer getCqlConditionSerializer(String columnConstraintClassName)
+        {
+            if (columnConstraintClassName.equals(CqlConstraintFunctionCondition.class.getName()))
+                return CqlConstraintFunctionCondition.serializer;
+            else if (columnConstraintClassName.equals(ConstraintScalarCondition.class.getName()))
+                return ConstraintScalarCondition.serializer;
```

Comment:
Not a fan of string matching. 
Can you thinking of an alternative? Can condition expose its own serializer directly?

Reply:
Thing is, it depends on what (who) serialized it in the first place. The name on this factory comes from
```
out.writeUTF(cqlConstraint.constraintCondition.getClass().toString());
```

in the serializer. I need to figure out which one to use when deserializing.